src/realm/sync/noinst/migration_store.cpp

```diff
@@ -0,0 +1,136 @@
+
+class MigrationStoreInit : public MigrationStore {
+public:
+    explicit MigrationStoreInit(DBRef db,
+                                std::function<void(MigrationStore::MigrationState)>&& on_migration_state_changed)
+        : MigrationStore(std::move(db), std::move(on_migration_state_changed))
+    {
+    }
+};
+
+} // namespace
+
+MigrationStoreRef
+MigrationStore::create(DBRef db, std::function<void(MigrationStore::MigrationState)>&& on_migration_state_changed)
+{
+    return std::make_shared<MigrationStoreInit>(std::move(db), std::move(on_migration_state_changed));
+}
+
+MigrationStore::MigrationStore(DBRef db,
+                               std::function<void(MigrationStore::MigrationState)>&& on_migration_state_changed)
+    : m_db(std::move(db))
+    , m_on_migration_state_changed(std::move(on_migration_state_changed))
+{
+    std::vector<SyncMetadataTable> internal_tables{
+        {&m_migration_table,
+         c_flx_migration_table,
+         {
+             {&m_migration_started_at, c_flx_migration_started_at, type_Timestamp},
+             {&m_migration_started_at, c_flx_migration_completed_at, type_Timestamp},
+             {&m_migration_state, c_flx_migration_state, type_Int},
+             {&m_migration_query_str, c_flx_migration_query_string, type_String},
+         }},
+    };
+
+    auto tr = m_db->start_read();
+    SyncMetadataSchemaVersions schema_versions(tr);
+
+    if (auto schema_version = schema_versions.get_version_for(tr, internal_schema_groups::c_flx_migration_store);
+        !schema_version) {
+        tr->promote_to_write();
+        schema_versions.set_version_for(tr, internal_schema_groups::c_flx_migration_store, c_schema_version);
+        create_sync_metadata_schema(tr, &internal_tables);
+        // create migration object
+        auto migration_store_obj = tr->get_table(m_migration_table)->create_object();
+        migration_store_obj.set(m_migration_state, MigrationState::NotStarted);
+        tr->commit_and_continue_as_read();
+        m_state = MigrationState::NotStarted;
+    }
+    else {
+        if (*schema_version != c_schema_version) {
+            throw std::runtime_error("Invalid schema version for flexible sync migration store metadata");
+        }
+        load_sync_metadata_schema(tr, &internal_tables);
+    }
+}
+
+std::shared_ptr<realm::SyncConfig> MigrationStore::convert_sync_config(std::shared_ptr<realm::SyncConfig> config)
+{
+    REALM_ASSERT(config);
+
+    std::lock_guard lock{m_mutex};
+    if (config->flx_sync_requested) {
+        cancel_migration();
+        return config;
+    }
+    if (m_state == MigrationState::NotStarted) {
+        return config;
+    }
+
+    auto flx_config = std::make_shared<realm::SyncConfig>(*config); // deep copy
+    flx_config->partition_value = "";
+    flx_config->flx_sync_requested = true;
+
+    return flx_config;
+}
+
+void MigrationStore::migrate_to_flx(std::string rql_query_string)
```

Comment:
Also add a "migration complete" function that just updates the state to `Completed` and calls `m_on_migration_state_changed()` with the new state.

Reply:
since the sync session would most likely have to call migration_complete, I find it a bit odd to call back with the new state (it's already known)

Reply:
maybe we should update it once we know how it will work

Reply:
since 'completed' it's supposed to be the ongoing state, I will use only two states: NotStarted and Completed (set when migrate_to_flx is called)